File: chap4/Python/exo1.py
import os

r1,w1=os.pipe() # fils->père
r2,w2=os.pipe() #père->fils

pid_fils=os.fork()

if pid_fils==0:
    os.close(r1)
    os.close(w2)
    mot=input("Entrer le mot :")
    os.write(w1,mot.encode()) #chaine.encode taille en bytes

    resultat_bytes=os.read(r2,4)
    resultat=int(resultat_bytes.decode()) #decode bytes->str

    if resultat==1:
        print("le mot est trouvé dans le fichier",resultat)
    else:
        print("le mot n'est pas trouvé dans le fichier",resultat)    
else:
   os.close(w1)
   os.close(r2)
   
   mot_bytes=os.read(r1,1024)
   mot=mot_bytes.decode()
   nombre=0
   fichier="test.txt"

with open(fichier, "r") as f:
    for ligne in f:
        for mot_fic in ligne.split():
            if mot == mot_fic:
              nombre =1

os.write(w2,str(nombre).encode())

    

File: chap4/C_/exo1.c
#include<stdio.h>
#include<stdlib.h>
#include<unistd.h>
#include<string.h>
#include<sys/wait.h>
/*Écrire un programme qui crée deux processus. Le processus père ouvre un ﬁchier texte en lecture. On suppose que le ﬁchier est composé de mots formés de
caractères alphabétiques séparés par des espaces. Le processus ﬁls saisit un mot au clavier. Le
processus père recherche le mot dans le ﬁchier, et transmet au ﬁls la valeur 1 si le mot est dans
le ﬁchier, et 0 sinon. Le ﬁls aﬃche le résultat.*/

int main()
{
    int tube1[2],tube2[2]; //tube 1 fils->père, tube2 père->fils
    char *fichier1="test.txt";
    char buffer[100],mot[100];
    int resultat=0,nombre,nombre2;
    FILE *fichier=fopen(fichier1, "r+");
    pipe(tube1);
    pipe(tube2);

    pid_t pid_fils=fork();

    if(pid_fils==0)
    {
        close(tube1[0]); //ne lit pas
        close(tube2[1]);//n'écrit pas
        printf("Entrer le mot que vous voulez tester si il est dans le fichier :");
        scanf("%s",mot);
        write(tube1[1],&mot,sizeof(mot));
        nombre=read(tube2[0],&resultat,sizeof(resultat));
        if(nombre>0)
        {
          printf(" résultat %d",resultat);
        }
        
    }
    else
    {
        close(tube1[1]); //n'écrit pas
        close(tube2[0]); //ne lit pas
        read(tube1[0],buffer,sizeof(buffer));
        nombre2=0;
        while(fscanf(fichier,"%s ",mot)!=EOF)
        {
            if(strcmp(mot,buffer)==0)
            {
                nombre2=1;
                break;
            }
           
        }
        write(tube2[1],&nombre2,sizeof(nombre2));

    }
   
}